File: server/protocol.go
// Package server WebSocket 訊息格式：登入、房間視野、依出口移動。傳統 MUD 節點連接節點。
package server

// ClientMsg 客戶端送出的 JSON；type 決定行為。
type ClientMsg struct {
	Type        string `json:"type"`         // "login" | "create_character" | "move" | "ping" | "get_entity_status" | "print_topology_debug"
	PlayerID    string `json:"player_id"`    // login 必填；create_character 時為新角色 id
	Password    string `json:"password"`     // login / create_character 必填
	DisplayChar string `json:"display_char"` // create_character 選填，預設「我」
	Gender      string `json:"gender"`       // create_character 選填：「男」→M、「女」→F，預設 M
	Direction   string `json:"direction"`    // move 時出口方向（例："東"、"西"）
	EntityID    string `json:"entity_id"`    // get_entity_status 時選填，空則為自己
}

// ViewEntity 房間內單一實體，供前端顯示「誰在這裡」。
type ViewEntity struct {
	ID          string `json:"id"`
	Kind        string `json:"kind"`
	DisplayChar string `json:"display_char"`
}

// ExitView 單一出口，供前端顯示可點選的出口。
type ExitView struct {
	Direction   string `json:"direction"`
	ToRoomID    string `json:"to_room_id"`
	ToRoomName  string `json:"to_room_name"`
}

// RoomViewMsg 伺服器推送：當前房間描述、出口列表、同房實體、遊戲時間。
type RoomViewMsg struct {
	Type                  string       `json:"type"`
	RoomID                string       `json:"room_id"`
	RoomName              string       `json:"room_name"`
	Description           string       `json:"description"`
	Exits                 []ExitView   `json:"exits"`
	Entities              []ViewEntity `json:"entities"`
	ServerUnix               int64  `json:"server_unix"`                     // 此 view 送出時的真實 Unix 秒，供前端插值
	GameTimeSecSinceMidnight int    `json:"game_time_sec_since_midnight"`   // 遊戲內當日 0:00 起的秒數（0～86399）
	GameDaysSinceEpoch       int    `json:"game_days_since_epoch"`           // 自 epoch 起算的遊戲日數，奇點曆用
}

// MeMsg 伺服器推送：登入成功後回傳玩家 id、當前房間、體敏氣與四項資源、命途/本源/星盤用欄位（狀態與星盤分頁規格 §五）。
type MeMsg struct {
	Type            string    `json:"type"`
	PlayerID        string    `json:"player_id"`
	RoomID          string    `json:"room_id"`
	RoomName        string    `json:"room_name"`
	Vit             int       `json:"vit"`
	Qi              int       `json:"qi"`
	Dex             int       `json:"dex"`
	HpCur           int       `json:"hp_cur"`
	HpMax           int       `json:"hp_max"`
	InnerCur        int       `json:"inner_cur"`
	InnerMax        int       `json:"inner_max"`
	SpiritCur       int       `json:"spirit_cur"`
	SpiritMax       int       `json:"spirit_max"`
	StaminaCur      int       `json:"stamina_cur"`
	StaminaMax      int       `json:"stamina_max"`
	DisplayTitle    string   `json:"display_title,omitempty"`    // 命途；空則前端顯示「無名之輩」
	OriginSentence  string   `json:"origin_sentence,omitempty"` // 本源一句話，後端拼句，三軸不傳
	ActivatedNodes  []string          `json:"activated_nodes,omitempty"` // 星盤已貫通節點 ID
	TopologyCosts   []float64         `json:"topology_costs,omitempty"` // 760 條 Cost（僅自己，供星盤綁定）
	EquipmentSlots  map[string]string `json:"equipment_slots,omitempty"` // slot→item_id
	EquipmentNames  map[string]string `json:"equipment_names,omitempty"` // slot→item_name（前端顯示用）
}

// MovedMsg 伺服器推送：某角色經出口移動後廣播給所有人。
type MovedMsg struct {
	Type      string `json:"type"`       // "moved"
	PlayerID  string `json:"player_id"`
	RoomID    string `json:"room_id"`
	RoomName  string `json:"room_name"`
}

// ErrorMsg 伺服器推送：錯誤說明。
type ErrorMsg struct {
	Type    string `json:"type"`    // "error"
	Message string `json:"message"`
}

// BlockedMsg 伺服器推送：無此出口或移動被阻擋。
type BlockedMsg struct {
	Type      string `json:"type"`       // "blocked"
	Direction string `json:"direction"`
}

// PongMsg 伺服器回覆心跳，供前景 keep-alive 用。
type PongMsg struct {
	Type string `json:"type"` // "pong"
}

// TopologyDebugAckMsg 暫時除錯用：print_topology_debug 已於伺服器終端印出，回傳此 ack 供前端得知。
type TopologyDebugAckMsg struct {
	Type    string `json:"type"`    // "topology_debug"
	Message string `json:"message"`
}

// EntityStatusMsg 伺服器回傳：單一實體狀態分頁用（體敏氣、四項資源、鎂；自己時含命途/本源/星盤）。
type EntityStatusMsg struct {
	Type            string    `json:"type"`
	EntityID        string    `json:"entity_id"`
	DisplayChar     string    `json:"display_char"`
	Vit             int       `json:"vit"`
	Qi              int       `json:"qi"`
	Dex             int       `json:"dex"`
	HpCur           int       `json:"hp_cur"`
	HpMax           int       `json:"hp_max"`
	InnerCur        int       `json:"inner_cur"`
	InnerMax        int       `json:"inner_max"`
	SpiritCur       int       `json:"spirit_cur"`
	SpiritMax       int       `json:"spirit_max"`
	StaminaCur      int       `json:"stamina_cur"`
	StaminaMax      int       `json:"stamina_max"`
	Magnesium       *int      `json:"magnesium"`
	IsSelf          bool      `json:"is_self"`
	DisplayTitle    string    `json:"display_title,omitempty"`
	OriginSentence  string    `json:"origin_sentence,omitempty"`
	ActivatedNodes  []string          `json:"activated_nodes,omitempty"`
	TopologyCosts   []float64         `json:"topology_costs,omitempty"`
	EquipmentSlots  map[string]string `json:"equipment_slots,omitempty"`
	EquipmentNames  map[string]string `json:"equipment_names,omitempty"`
}

File: src/server/protocol.rs
//! WebSocket／HTTP JSON 訊息型別（對齊既有 `server/protocol`）。
//! 欄位名與 `omitempty` 行為需與前端一致。

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// 客戶端送出的 JSON；`type` 決定行為。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ClientMsg {
    #[serde(rename = "type")]
    pub msg_type: String,
    #[serde(default)]
    pub player_id: String,
    #[serde(default)]
    pub password: String,
    #[serde(default, rename = "display_char")]
    pub display_char: String,
    #[serde(default)]
    pub gender: String,
    #[serde(default)]
    pub direction: String,
    #[serde(default, rename = "entity_id")]
    pub entity_id: String,
    #[serde(default, rename = "item_id")]
    pub item_id: String,
    #[serde(default, rename = "target_slot")]
    pub target_slot: String,
    #[serde(default)]
    pub slot: String,
    #[serde(default)]
    pub action: String,
    #[serde(default, rename = "player_input")]
    pub player_input: String,
}

/// 房間內單一實體（對齊 `ViewEntity`）。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ViewEntity {
    pub id: String,
    pub kind: String,
    #[serde(rename = "display_char")]
    pub display_char: String,
    #[serde(rename = "display_name")]
    pub display_name: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub actions: Vec<String>,
}

/// `do_action` 敘事結果（對齊 `ActionResultMsg`）。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActionResultMsg {
    #[serde(rename = "type")]
    pub msg_type: String,
    pub action: String,
    #[serde(rename = "target_id")]
    pub target_id: String,
    #[serde(rename = "target_name")]
    pub target_name: String,
    pub narrative: String,
    pub success: bool,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub actions: Vec<String>,
    #[serde(
        default,
        skip_serializing_if = "String::is_empty",
        rename = "move_target_id"
    )]
    pub move_target_id: String,
}

/// 單一出口（對齊 `ExitView`）。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ExitView {
    pub direction: String,
    #[serde(rename = "to_room_id")]
    pub to_room_id: String,
    #[serde(rename = "to_room_name")]
    pub to_room_name: String,
}

/// 房間內可互動物件（對齊 `ViewObject`）。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ViewObject {
    pub id: String,
    pub name: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub actions: Vec<String>,
}

/// 當前房間視野（對齊 `RoomViewMsg`）。
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RoomViewMsg {
    #[serde(rename = "type")]
    pub msg_type: String,
    #[serde(rename = "room_id")]
    pub room_id: String,
    #[serde(rename = "room_name")]
    pub room_name: String,
    pub description: String,
    pub exits: Vec<ExitView>,
    pub entities: Vec<ViewEntity>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub objects: Vec<ViewObject>,
    #[serde(rename = "server_unix")]
    pub server_unix: i64,
    #[serde(rename = "game_time_sec_since_midnight")]
    pub game_time_sec_since_midnight: i32,
    #[serde(rename = "game_days_since_epoch")]
    pub game_days_since_epoch: i32,
}

/// 登入成功後玩家狀態（對齊 `MeMsg`）。
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MeMsg {
    #[serde(rename = "type")]
    pub msg_type: String,
    #[serde(rename = "player_id")]
    pub player_id: String,
    #[serde(rename = "room_id")]
    pub room_id: String,
    #[serde(rename = "room_name")]
    pub room_name: String,
    pub vit: i32,
    pub qi: i32,
    pub dex: i32,
    #[serde(rename = "hp_cur")]
    pub hp_cur: i32,
    #[serde(rename = "hp_max")]
    pub hp_max: i32,
    #[serde(rename = "inner_cur")]
    pub inner_cur: i32,
    #[serde(rename = "inner_max")]
    pub inner_max: i32,
    #[serde(rename = "spirit_cur")]
    pub spirit_cur: i32,
    #[serde(rename = "spirit_max")]
    pub spirit_max: i32,
    #[serde(rename = "stamina_cur")]
    pub stamina_cur: i32,
    #[serde(rename = "stamina_max")]
    pub stamina_max: i32,
    #[serde(
        default,
        skip_serializing_if = "String::is_empty",
        rename = "display_title"
    )]
    pub display_title: String,
    #[serde(
        default,
        skip_serializing_if = "String::is_empty",
        rename = "origin_sentence"
    )]
    pub origin_sentence: String,
    #[serde(
        default,
        skip_serializing_if = "Vec::is_empty",
        rename = "activated_nodes"
    )]
    pub activated_nodes: Vec<String>,
    #[serde(
        default,
        skip_serializing_if = "Vec::is_empty",
        rename = "topology_costs"
    )]
    pub topology_costs: Vec<f64>,
    #[serde(
        default,
        skip_serializing_if = "HashMap::is_empty",
        rename = "equipment_slots"
    )]
    pub equipment_slots: HashMap<String, String>,
    #[serde(
        default,
        skip_serializing_if = "HashMap::is_empty",
        rename = "equipment_names"
    )]
    pub equipment_names: HashMap<String, String>,
}

/// 移動廣播（對齊 `MovedMsg`）。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MovedMsg {
    #[serde(rename = "type")]
    pub msg_type: String,
    #[serde(rename = "player_id")]
    pub player_id: String,
    #[serde(rename = "room_id")]
    pub room_id: String,
    #[serde(rename = "room_name")]
    pub room_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ErrorMsg {
    #[serde(rename = "type")]
    pub msg_type: String,
    pub message: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct BlockedMsg {
    #[serde(rename = "type")]
    pub msg_type: String,
    pub direction: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct PongMsg {
    #[serde(rename = "type")]
    pub msg_type: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct NarrateMsg {
    #[serde(rename = "type")]
    pub msg_type: String,
    pub text: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TopologyDebugAckMsg {
    #[serde(rename = "type")]
    pub msg_type: String,
    pub message: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct InventoryItemView {
    #[serde(rename = "item_id")]
    pub item_id: String,
    pub name: String,
    #[serde(rename = "item_type")]
    pub item_type: String,
    pub qty: i32,
    pub weight: f64,
    #[serde(rename = "sub_total")]
    pub sub_total: f64,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub description: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub slot: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct InventoryMsg {
    #[serde(rename = "type")]
    pub msg_type: String,
    pub items: Vec<InventoryItemView>,
    #[serde(rename = "current_weight")]
    pub current_weight: f64,
    #[serde(rename = "max_weight")]
    pub max_weight: f64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct EntityStatusMsg {
    #[serde(rename = "type")]
    pub msg_type: String,
    #[serde(rename = "entity_id")]
    pub entity_id: String,
    #[serde(rename = "display_char")]
    pub display_char: String,
    pub vit: i32,
    pub qi: i32,
    pub dex: i32,
    #[serde(rename = "hp_cur")]
    pub hp_cur: i32,
    #[serde(rename = "hp_max")]
    pub hp_max: i32,
    #[serde(rename = "inner_cur")]
    pub inner_cur: i32,
    #[serde(rename = "inner_max")]
    pub inner_max: i32,
    #[serde(rename = "spirit_cur")]
    pub spirit_cur: i32,
    #[serde(rename = "spirit_max")]
    pub spirit_max: i32,
    #[serde(rename = "stamina_cur")]
    pub stamina_cur: i32,
    #[serde(rename = "stamina_max")]
    pub stamina_max: i32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub magnesium: Option<i32>,
    #[serde(rename = "is_self")]
    pub is_self: bool,
    #[serde(
        default,
        skip_serializing_if = "String::is_empty",
        rename = "display_title"
    )]
    pub display_title: String,
    #[serde(
        default,
        skip_serializing_if = "String::is_empty",
        rename = "origin_sentence"
    )]
    pub origin_sentence: String,
    #[serde(
        default,
        skip_serializing_if = "Vec::is_empty",
        rename = "activated_nodes"
    )]
    pub activated_nodes: Vec<String>,
    #[serde(
        default,
        skip_serializing_if = "Vec::is_empty",
        rename = "topology_costs"
    )]
    pub topology_costs: Vec<f64>,
    #[serde(
        default,
        skip_serializing_if = "HashMap::is_empty",
        rename = "equipment_slots"
    )]
    pub equipment_slots: HashMap<String, String>,
    #[serde(
        default,
        skip_serializing_if = "HashMap::is_empty",
        rename = "equipment_names"
    )]
    pub equipment_names: HashMap<String, String>,
    #[serde(
        default,
        skip_serializing_if = "HashMap::is_empty",
        rename = "equipment_descs"
    )]
    pub equipment_descs: HashMap<String, String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn client_msg_minimal_json() {
        let j = r#"{"type":"ping"}"#;
        let m: ClientMsg = serde_json::from_str(j).unwrap();
        assert_eq!(m.msg_type, "ping");
        assert!(m.player_id.is_empty());
    }

    #[test]
    fn room_view_roundtrip() {
        let v = RoomViewMsg {
            msg_type: "view".into(),
            room_id: "r1".into(),
            room_name: "大廳".into(),
            description: "…".into(),
            exits: vec![ExitView {
                direction: "東".into(),
                to_room_id: "r2".into(),
                to_room_name: "走廊".into(),
            }],
            entities: vec![ViewEntity {
                id: "p1".into(),
                kind: "player".into(),
                display_char: "我".into(),
                display_name: "p1".into(),
                actions: vec![],
            }],
            objects: vec![],
            server_unix: 1700000000,
            game_time_sec_since_midnight: 3600,
            game_days_since_epoch: 1,
        };
        let s = serde_json::to_string(&v).unwrap();
        let v2: RoomViewMsg = serde_json::from_str(&s).unwrap();
        assert_eq!(v, v2);
    }

    #[test]
    fn entity_status_magnesium_omit() {
        let e = EntityStatusMsg {
            msg_type: "entity_status".into(),
            entity_id: "n1".into(),
            display_char: "店".into(),
            vit: 1,
            qi: 2,
            dex: 3,
            hp_cur: 10,
            hp_max: 10,
            inner_cur: 0,
            inner_max: 0,
            spirit_cur: 0,
            spirit_max: 0,
            stamina_cur: 0,
            stamina_max: 0,
            magnesium: None,
            is_self: false,
            display_title: String::new(),
            origin_sentence: String::new(),
            activated_nodes: vec![],
            topology_costs: vec![],
            equipment_slots: HashMap::new(),
            equipment_names: HashMap::new(),
            equipment_descs: HashMap::new(),
        };
        let s = serde_json::to_string(&e).unwrap();
        assert!(!s.contains("magnesium"));
        let e2: EntityStatusMsg = serde_json::from_str(&s).unwrap();
        assert_eq!(e2.magnesium, None);
    }

    #[test]
    fn action_result_with_move_target_id() {
        let j = r#"{"type":"action_result","action":"Look","target_id":"o1","target_name":"門","narrative":"…","success":true,"move_target_id":"door_1"}"#;
        let a: ActionResultMsg = serde_json::from_str(j).unwrap();
        assert_eq!(a.move_target_id, "door_1");
        let out = serde_json::to_string(&a).unwrap();
        assert!(out.contains("move_target_id"));
    }
}
